src/bin/lychee/main.rs

```diff
@@ -184,6 +231,68 @@ async fn run(cfg: &Config, inputs: Vec<Input>) -> Result<i32> {
     }
 }
 
+async fn recurse(
+    response: Response,
+    cache: &mut Cache,
+    input_domains: &HashSet<String>,
+    cfg: &Config,
+    pb: &Option<ProgressBar>,
+    send_req: Sender<Request>,
+) -> Result<usize> {
+    let recursion_level = response.recursion_level + 1;
+
+    if let Some(depth) = cfg.depth {
+        if recursion_level > depth {
+            // Maximum recursion depth reached; stop link checking.
+            return Ok(0);
+        }
+    }
+
+    if !response.status.is_success() {
+        return Ok(0);
+    }
+    if cache.contains(response.uri.as_str()) {
+        return Ok(0);
+    }
+    cache.insert(response.uri.to_string());
```

Comment:
Since we're allowing *infinite* recursion, this data structure can really take up the entire memory with sufficient number of links. Depending on the expected maximum number of links checked in recursive mode, we might consider a [bloom filter](https://en.wikipedia.org/wiki/Bloom_filter) here:
- HashSet is probably fine if the number of links checked is in the order of millions or fewer
- Bloom filter might be a good idea if the number of links checked is in the order of tens of millions or more

---

Super-rough calculations below. (Note that I'm just calculating the size of this one structure, there's probably going to be more memory overheads from various bits of the program).

Assumptions:
- 64 bit machine, e.g. 8 bytes per pointer
- 64 bytes per link on average
- 1 million links

Calculations:

- Link string storage: 64 MB
- String overhead (24 bytes per string): 24 MB
  source: https://cheats.rs/#owned-strings
- HashSet overhead is 1 byte per entry: 1MB
  source: https://github.com/rust-lang/hashbrown#features

So, HashSet with 1 million unique links will take up ~64+24+1 = ~89 MB (~85 MiB)

For comparison, a Bloom Filter with ~1e-7 false positive probability is expected to take up:
- ~4MiB for 1 million links: https://hur.st/bloomfilter/?n=1M&p=1.0E-7&m=&k=
- ~80MiB for 20 million liks: https://hur.st/bloomfilter/?n=20M&p=1.0E-7&m=&k=
- ~4GiB for 1 billion links: https://hur.st/bloomfilter/?n=1G&p=1.0E-7&m=&k=

So, to be honest, HashSet should be absolutely fine for the most users. But we should make a remark in the README that memory usage when using recursive mode with tens millions of links might cause RAM usage issues.

Reply:
Hi, would it be merged soon? and Docker image updated with the change to make GH action use the latest?

Reply:
@pawroman: Good idea with the bloom filter. I'll add that as soon as #208 is merged.
@lenisha: Hopefully yes. I don't like the current implementation anymore (see comment below), but I think it's not gonna be a blocker either way.